File: p1.py
#Time: O(N)
#space: O(1)
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if nums is None or len(nums)==0:
            return False
        n = len(nums)
        reachable = 0 
        for i in range(n):
            if i > reachable:
                return False
            reachable = max(reachable,nums[i]+i)
        
        return True
            
        

File: p1.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

class Solution {
    public boolean canJump(int[] nums) {
        int target = nums.length - 1;
        ////check if from the current element, we can reach the target, if yes current element would be the new target
        for(int i = nums.length-2; i>=0; i--){
            if(nums[i] + i >= target){
                target = i;
            }
        }

        return target == 0;
    }
}